<img src="../Data/images/ZumiHeader.png" width=700>

# Robot Emotions(機械人的情感)

<font size =3> Zumi是很有個性的！在這堂中，您將會學習Zumi是如果偵測人類的情感及如何編程Zumi的個性。您還將了解聲音、它的測量方式以及它與情感的對應關係。 最後您將會訓練Zumi去辨認並對它最愛的顏色作出反應。</font> 



## How do we detect emotion?(我們如何偵測情感？)
<font size =3> 看看下面的圖片，你能能識別出每種情感嗎？

<img src="../Data/images/emotions.png" width=700> <br>

您是怎樣判定那種情感是那種情感？有許多特徵可作為指標，例如：眼睛、嘴巴、眉毛，什至手勢。我們要怎樣做才能將情感轉化成機械人能理解的東西？
如果您有看過一套叫*Cars*的電影，您可能會知道每一部車都有它們自己的個性。車是如何表達情感的？透過動作？聲音？眼睛？</font> 


### Import libraries(導入庫)
<font size =3> 要用個性的函數，我們需要導入Zumi,螢幕及個性庫。</font> 


In [ ]:
from zumi.zumi import Zumi
from zumi.util.camera import Camera
from zumi.util.screen import Screen
from zumi.personality import Personality
from zumi.util.color_classifier import ColorClassifier
import time


zumi = Zumi()
screen = Screen()
personality = Personality(zumi, screen)
camera = Camera()

Starting Zumi 
Pi Zero I2C is available
Verified Pi Zero is the same
Gyroscope previously calibrated
Zumi board detected
Compass detected
OLED Screen detected
Gyroscope & Accelerometer detected


###  Calling personality functions(使用個性函數)
<font size =3> 以下是一些您能使用的：

* happy()
     
* celebrate()
       
* angry()
       
* look_around()

* look_around_open()
       
* disoriented_left()
       
* disoriented_right()

* awake()

例如：
<font face="Courier">personality.happy()</font> 會使Zumi擺動並發出聲音！
            
在下面的單元格中，測試其他個性函數並看看它們的作用。</font>

In [ ]:
personality.awake()

## Sounds(聲音)

<font size =3>Zumi能夠發出聲音以配合它的情感！聲音能以頻率(frequency)和振幅(amplitude)來測量。

*   Frequency(頻頻)是每秒振動數或每秒脈衝數(pulses)，均以hertz(赫茲)為單位。 Frequency(頻頻)愈高，聲音的音調便愈高。
*   Amplitude(振幅) 是聲音有多大或多強，以decibels(分貝)為單位。Amplitude(振幅)愈高，聲音愈大。 

影片: [Sound: Wavelength, Frequency, and Amplitude](https://www.youtube.com/watch?v=TsQL-sXZOLc)

每一種的情感的聲音是怎麼的？開心是高還是低頻率(frequency)？憤怒的振幅(amplitude)是高還是低？要如何應用在Zumi上呢？

您能使用<font face="Courier">play_note()</font>來彈奏各種音符。第一個參數是您要演奏的音符(C2至B6的任何一個音符)。第二個參數是非必要的，是您希望音符播放的時間（（以毫秒為單位）。預設的數值為500ms，但您能透過利用下面的函數來改變它： <br><br>
<font face="Courier">play_note(Note.GS3, 400)</font>. <br><br>
這會播放note G Sharp below middle C,持續4ms。試試以下的編程，並嘗試創作自己的音樂！</font>


In [ ]:
from zumi.protocol import Note 
zumi.play_note(Note.C4)
zumi.play_note(Note.D4)
zumi.play_note(Note.E4)
zumi.play_note(Note.F4)
zumi.play_note(Note.G4)
zumi.play_note(Note.A4)
zumi.play_note(Note.B4)
zumi.play_note(Note.C5)

<font size =3> 編寫屬於您的開心，傷心，憤怒或興奮的聲音。嘗試使用不同旋律創作出您的音樂。</font>

In [ ]:
# Make your melodies here 🎵 

## Screen(螢幕)

<font size =3> Zumi亦能使用 **OLED** (organic LED) screen來表達情感。

Zumi有不同的「眼睛」:

* <font face="Courier"> close_eyes()</font>
* <font face="Courier"> sleepy_eyes()</font>
* <font face="Courier"> sleepy_left()</font>
* <font face="Courier"> sleepy_right()</font>
* <font face="Courier"> blink()</font>
* <font face="Courier"> look_around_open()</font>
* <font face="Courier"> sleeping()</font>
* <font face="Courier"> look_around()</font>     
* <font face="Courier"> glimmer()</font>
* <font face="Courier"> sad()</font>
* <font face="Courier"> happy()</font>
* <font face="Courier"> hello()</font>
* <font face="Courier"> angry()</font>

要使用螢幕，請使用您要的screen class(螢幕種類)，試用: </font>

In [ ]:
screen.close_eyes()

NameError: name 'screen' is not defined

### Draw Text(繪製文本)

<font size =3> 除了繪畫Zumi的眼睛外，您還能在螢幕上繪寫訊息！例如利用 <font face="Courier">draw_text()</font>來繪寫訊息: </font>

In [ ]:
screen.draw_text("hello!")

<font size =3> 如果您想把文字自動置中可使用： </font>

In [ ]:
screen.draw_text_center("hello!")

<font size =3> 如果您想使用數字，您需要使用<font face="Courier">String</font>來確保正使用數據格式。</font>

In [ ]:
number = 10
screen.draw_text("ten " + str(number)) # the str() functions turns the number into a string

<font size =3> 您什至能讓Zumi顯示時間！</font>

In [ ]:
for i in range(0,50):
    screen.draw_text_center(time.ctime())
    time.sleep(0.1)

## Eyes on Me(看著我)
<font size = 3> 在這個活動中,Zumi會一直看著您！您需編寫一個讓Zumi的眼睛跟隨著您的樣子移動的程式。如果Zumi看不到您，它會變得傷心！
這個編程跟課堂4.2 Face Detction的非常相似，如果忘記的話可以回去參考。 </font>
    
### Dividing the frame into left, right, and middle(將框架分為左、右、中)
<font size=3> 第一步我們需要知道Zumi相機的像素是多少。要這樣做的話，我們可以這樣查看相機拍攝對象的屬性。

In [ ]:
print(camera.image_width)

160


<font size=3> 在預設下，像素寬度為160。雖然您能改變它,但我們先會這個數值作例子。現在，將圖像框架劃分為左側、右側和中間區域。您可以平分或不平分也可以，都是看你的選擇。但請記緊您只有160pixels去分成三個部份，盡量不要有一部份佔據多於一半的像素量。

## You Smile, I Smile(您微笑時我也微笑)

<font size=3>您已經學會利用聲音跟螢幕來使Zumi人性化了，現在讓它對微笑做出反應！在上一課中您使用過<font face="Courier">vision.find_face()</font>來偵測人臉。您覺得笑臉的特徵是怎樣的？嘴巴上較暗的像素會是向上還是向下傾斜？在這節，您將使用一個函數來檢查Zumi偵測到的人臉是否在笑。如果Zumi看到笑容，它便會表現得高興。如果沒有，Zumi便會的眼睛會變得傷心。</font>
    
### vision.find_smile()
<font size=3> 您將用到的函數叫作<font face="Courier">vision.find_smile()</font>. 當Zumi偵測到人臉時，這個函數會變回像素座標(pixel coordinates)，如果不是笑臉便會變成<font face="Courier">(-1,-1,-1,-1)</font>例如，當您拍照和儲存相片後，您可以讓Zumi會偵查相片中有沒有笑臉：<br> <br>
    
<font face="Courier">
    
(x,y,w,h) = vision.find_smile(image)<br>
if (x,y,w,h) == (-1,-1,-1,-1): <br>
<span style="margin-left: 40px;"># Zumi did not find a smile :(</span> <br>
else: <br>
<span style="margin-left: 40px;"># Zumi sees a smile! :)</span> <br>
    </font> </font>
    
### Code It!(試編寫它！)
<font size=3> 試編些一個讓Zumi拍照並尋找笑臉的程式。 <br><br>**Hint:** 緊記在Zumi偵測不到笑臉時改變參數。填寫以下編程所缺的地方：

In [ ]:
camera.start_camera()
try:
    for i in range(1000):
        # Finish the code!

        
finally:
    camera.close()

<hr>

# Extension Activities(延伸活動)<br> 

<img src="../Data/images/physics_extension.jpg" width=70 align="left">

###  Frequency(頻率) <br> 
<font size =3>使用調諧器(tuners)去識別Zumi發出不同的聲音的頻率、波長和幅度。 那個情感有較高的頻率、波長和幅度？ </font><br><br>


### Add personality to color classifier lesson!(在顏色分類器課堂中添加個性！)
<font size =3> <span style="color:red"> **Note!** </span> 這個活動需要在Zumi dashboard中的"Explore"使用color Training Wizard。<br><br>
回到Color Training Wizard並訓練Zumi學習各種顏色。加載下面的模型，替Zumi設定它最愛的顏色，當偵查到最愛的顏色時便做出開心的反應，而當偵測到其他顏色時便做出憤怒反應！讓同學們向Zumi展示顏色卡並猜測Zumi最愛的顏色是什麼！</font> 

In [ ]:
camera = Camera()
knn = ColorClassifier()
train = knn.load_model("PROJECT NAME HERE")
knn.fit("hsv")

camera.start_camera()

while True:
        user_input = input("Press 'enter' to predict or 'q to quit: ")

        if user_input == "q":
            break
            
        image = camera.capture()
        predict = knn.predict(image)
        screen.draw_text_center(predict)
        
        # Add your if statements here!

camera.close()